Implement `encoding` in .rules

jokesper opened the pull request:
<!--
Thanks for your pull request! We appreciate it. 
If you're a new developer, FOSS contributor, or hledger contributor, welcome.

Much of our best design work and knowledge sharing happens during code review,
so be prepared for more work ahead, especially if your PR is large.
To minimise waste, and get your hledger PRs accepted quickly,
please check the guidelines in the developer docs:

https://hledger.org/PULLREQUESTS.html
https://hledger.org/COMMITS.html

You don't need to master our commit conventions, but do add at least one prefix and colon
to the commit message summary. The most common prefixes are:

- feat: for user-visible features
- fix:  for user-visible fixes
- imp:  for user-visible improvements
- ;doc:  for documentation improvements  (; at the start enables quicker/cheaper CI tests)
- dev:  for internal improvements

-->
An implementation for `encoding` in rules file:
The cases are:
- [X] `-f *.rules`
- [x] `-f *.csv`
- [x] `-f *.csv --rules *.rules`

The second and third case are effectively the same. The problem is that for them we call a `Reader`s `rReadFn` value taking the contents rather then the filepath / handle.
So we either would have to change `Reader` to take a `FilePath` instead of `Text`, which would break how we do Tests,
or we add an extra field  `rGetContents :: FilePath -> IO Text`.
I would rather get some feedback on this before I implement a potentially bigger change.

If implemented this PR should:
- fix #1777

jokesper committed:
feat: Implement `encoding` in rules file

Instead of `text-icu` as [recommended by `text`](https://hackage.haskell.org/package/text-2.1.2/docs/Data-Text-Encoding.html):

> To gain access to a much larger family of encodings, use the `text-icu` package.

we use `encoding`, since `text-icu` requires an external library.
`encoding` does require a custom setup, which is currently
affected by a [bug in Cabal](https://github.com/haskell/cabal/issues/6505)
but it doesn't require an external runtime dependency, which makes it
easier to distribute on macos and windows.

I also changed the `rReadFn` to get a `Handle` rather than a `Text`.
Such that we can use non standard functions for reading it.
This is kind of sideways, since the one place we care about it,
we don't use it at all. Though it still caused issues, since the input
wasn't parseable as text.
I think not using the handle has something to do with how `-` is treated
as stdin but I didn't do any further tests on that, since this is out of
the scope of this PR.

simonmichael commented:
Thanks for exploring this. It's probably a good idea. But what if rules files made it easier to run a preprocessor, eg

    source !iconv -f CP1250 FILE.csv

That requires a little more expertise from the user, I guess ?

jokesper commented:
> But what if rules files made it easier to run a preprocessor

I like the idea, though implementing it with the other cases is a bit wierd, since we ignore `source` with `-f`.

We could split it into two options with the other being `preprocessor iconv -f CP1250 {}`.
This has the drawback of either reserving a replacement or allowing one to specify it.

> The case of reading CSV from stdin might also be worth thinking about.

Is it even possible to read a csv from stdin?
- `-f - --rules *.rules` doesn't work, since we first figure out whether it's a csv based on the file ending.
- `source -` doesn't work, since it searches for `-` as a file and can't find it.
- `source /dev/stdin` works, but is a bit hacky.

Though if they would work, they would be handled the same as the cases mentioned above.
Opening the handle and reading from it are indpendent.

simonmichael commented:
> Is it even possible to read a csv from stdin?
> 

Like so: `-f csv:- --rules-file=foo.rules`

simonmichael commented:
But it can run x86_64 ghc. And yes, it's similar on Windows: text-icu isn't an easy thing to install.
```
C:\Users\Simon>stack install text-icu
[1 of 3] Compiling Main             ( C:\sr\setup-exe-src\setup-O_vy6YIf.hs, C:\sr\setup-exe-src\setup-O_vy6YIf.o )
[2 of 3] Compiling StackSetupShim   ( C:\sr\setup-exe-src\setup-shim-O_vy6YIf.hs, C:\sr\setup-exe-src\setup-shim-O_vy6YIf.o )
[3 of 3] Linking C:\sr\setup-exe-cache\x86_64-windows\tmp-Cabal-simple_O_vy6YIf_3.10.3.0_ghc-9.8.4.exe
text-icu> configure
text-icu> Configuring text-icu-0.8.0.5...
text-icu> Error: Cabal-simple_O_vy6YIf_3.10.3.0_ghc-9.8.4.exe: The program 'pkg-config'
text-icu> version >=0.9.0 is required but it could not be found.
text-icu>

Error: [S-7282]
       Stack failed to execute the build plan.

       While executing the build plan, Stack encountered the error:

       [S-7011]
       While building package text-icu-0.8.0.5 (scroll up to its section to see the error) using:
       C:\sr\setup-exe-cache\x86_64-windows\Cabal-simple_O_vy6YIf_3.10.3.0_ghc-9.8.4.exe --verbose=1 --builddir=.stack-work\dist\f24b2e15 configure --with-ghc=C:\Users\Simon\AppData\Local\Programs\stack\x86_64-windows\ghc-9.8.4\bin\ghc-9.8.4.exe --with-ghc-pkg=C:\Users\Simon\AppData\Local\Programs\stack\x86_64-windows\ghc-9.8.4\bin\ghc-pkg-9.8.4.exe --user --package-db=clear --package-db=global --package-db=C:\sr\snapshots\ce0f12f1\pkgdb --libdir=C:\sr\snapshots\ce0f12f1\lib --bindir=C:\sr\snapshots\ce0f12f1\bin --datadir=C:\sr\snapshots\ce0f12f1\share --libexecdir=C:\sr\snapshots\ce0f12f1\libexec --sysconfdir=C:\sr\snapshots\ce0f12f1\etc --docdir=C:\sr\snapshots\ce0f12f1\doc\text-icu-0.8.0.5 --htmldir=C:\sr\snapshots\ce0f12f1\doc\text-icu-0.8.0.5 --haddockdir=C:\sr\snapshots\ce0f12f1\doc\text-icu-0.8.0.5 --dependency=base=base-4.19.2.0-943b --dependency=bytestring=bytestring-0.12.1.0-290b --dependency=deepseq=deepseq-1.5.1.0-f918 --dependency=text=text-2.1.1-3967 --dependency=time=time-1.12.2-0f79 -fhomebrew --extra-include-dirs=C:\Users\Simon\AppData\Local\Programs\stack\x86_64-windows\msys2-20240727\mingw64\include --extra-lib-dirs=C:\Users\Simon\AppData\Local\Programs\stack\x86_64-windows\msys2-20240727\mingw64\lib --extra-lib-dirs=C:\Users\Simon\AppData\Local\Programs\stack\x86_64-windows\msys2-20240727\mingw64\bin --exact-configuration --ghc-option=-fhide-source-paths
       Process exited with code: ExitFailure 1
```

jokesper commented:
Should I instead switch to something like [`encoding`](https://hackage.haskell.org/package/encoding)?
It would be easier to install, as it's installed entirely by `cabal` / `stack` and would be statically build into `hledger`.
The drawback is: It works on `String` not `Text`. I.e. we would need a possibly costly `pack` (not tested).

simonmichael commented:
Yes that looks pretty good. Converting from Text to String won't affect anyone noticeably I predict, though if it can be avoided when unnecessary, all the better.

jokesper commented:
I just switched the part I already implemented to use `encoding` instead of `text-icu`.

I noticed `encoding` uses a custom setup which is currently affected by [this bug](https://github.com/haskell/cabal/issues/6505). For some developers (on Arch btw.) this leads to an extra option they should pass: `--ghc-options=-dynamic` when using cabal.

Should I document this anywhere, as it's already documented in the Arch wiki.

jokesper commented:
I implemented the remaining two cases. To do this I had to change the arguments to `rReadFn`. The details on why are in the commit message.

I think this is ready for review.

jokesper committed:
;doc: Document new `encoding` option

jokesper commented:
> Docs and tests will help motivate this, when the time comes.

Writing tests for this would require quite a bit of rework in how they are treated.
Currently the readers for `csv` and `rules` only look at the filepath and not at their passed contents.
Instead they get them themselves.

simonmichael commented:
Hi @jokesper. Sorry for the slow reply. If this would make hledger installers need to use special build flags, that would be a problem unfortunately.

simonmichael commented:
I have inquired at https://github.com/dmwit/encoding/issues/23

jokesper commented:
While trying to fix the CI, I noticed you already needed to supply `--ghc-options=-dynamic` to build `hledger-web`, since it depends on `xml-conduit` which has a custom setup.
Furthermore this is only a problem with cabal (stack should work fine) and on the first build (since then the dependencies are build).

Though I would still prefer if we didn't need a flag just to build the core `hledger-lib`, which would also affect every project depending on `hledger`.

jokesper committed:
dev: Upgraded `encoding` dependency

This makes sure we don't have to build a `build-type: Custom` package

jokesper commented:
Since `encoding` now dropped the custom build type, the associated problems have now been fixed

simonmichael commented:
Thanks @jokesper . Adding encoding-0.9 to stack*.yaml should fix the current CI failure.

simonmichael commented:
(Or in the case of stack.yaml, bumping it to the latest nightly-2025-02-11.)

jokesper committed:
dev: fix ci

jokesper commented:
Thanks for the info. It's fixed now

simonmichael commented:
Thanks @jokesper.
Now it's time for more review, and I'd welcome help from other reviewers.
Here's some relevant info:

26 files changed

encoding >0.9 dependency added

## Main commit description

    feat: Implement encoding in rules file

    Instead of `text-icu` as [recommended by `text`](https://hackage.haskell.org/package/text-2.1.2/docs/Data-Text-Encoding.html):
    
    > To gain access to a much larger family of encodings, use the `text-icu` package.
    
    we use `encoding`, since `text-icu` requires an external library.
    `encoding` does require a custom setup, which is currently
    affected by a [bug in Cabal](https://github.com/haskell/cabal/issues/6505)
    but it doesn't require an external runtime dependency, which makes it
    easier to distribute on macos and windows.
    
    I also changed the `rReadFn` to get a `Handle` rather than a `Text`.
    Such that we can use non standard functions for reading it.
    This is kind of sideways, since the one place we care about it,
    we don't use it at all. Though it still caused issues, since the input
    wasn't parseable as text.
    I think not using the handle has something to do with how `-` is treated
    as stdin but I didn't do any further tests on that, since this is out of
    the scope of this PR.


## Github AI summary

This pull request includes several changes to the `hledger-lib` library, primarily focusing on modifying the journal reading functions to use `Handle` instead of `Text` and adding support for handling different text encodings. The changes span multiple files and involve updates to function signatures, imports, and helper functions.

Updates to journal reading functions:
* Changed `readJournal` and related functions to accept a `Handle` instead of `Text` for reading journal data (`hledger-lib/Hledger/Read.hs`). [[1]](diffhunk://#diff-bdf99fb149929f11680d239ef2f7508759abea3b651393dc6806a31e1e71e511L222-R227) [[2]](diffhunk://#diff-bdf99fb149929f11680d239ef2f7508759abea3b651393dc6806a31e1e71e511L316-R324)
* Added a new function `readJournal''` that allows reading from `Text` by converting it to a `Handle` (`hledger-lib/Hledger/Read.hs`).

Import and helper function updates:
* Updated imports to include `Handle` and other necessary modules (`hledger-lib/Hledger/Read.hs`, `hledger-lib/Hledger/Read/Common.hs`, `hledger-lib/Hledger/Read/CsvReader.hs`, `hledger-lib/Hledger/Read/RulesReader.hs`). [[1]](diffhunk://#diff-bdf99fb149929f11680d239ef2f7508759abea3b651393dc6806a31e1e71e511L148-R149) [[2]](diffhunk://#diff-fcee3c006b1d2e35b8c279c9daa9963ecdb5910bdc90099f7b3a37b8b529a65dR152) [[3]](diffhunk://#diff-52f2974c62ce69a30e2bddf4d8b6008441715fde0a9d0a4c3ea8c072fa465223L31-R31) [[4]](diffhunk://#diff-95e0d5df4894ae4a99b5ad4a37811c64ec30d82af3b01035561976afc3c4293eR73)
* Added helper functions such as `handleReadFnToTextReadFn` to facilitate the transition from `Text` to `Handle` (`hledger-lib/Hledger/Read/Common.hs`).

Changes in specific readers:
* Updated `CsvReader`, `JournalReader`, `RulesReader`, `TimeclockReader`, and `TimedotReader` to use the new `Handle`-based functions (`hledger-lib/Hledger/Read/CsvReader.hs`, `hledger-lib/Hledger/Read/JournalReader.hs`, `hledger-lib/Hledger/Read/RulesReader.hs`, `hledger-lib/Hledger/Read/TimeclockReader.hs`, `hledger-lib/Hledger/Read/TimedotReader.hs`). [[1]](diffhunk://#diff-52f2974c62ce69a30e2bddf4d8b6008441715fde0a9d0a4c3ea8c072fa465223L57-R68) [[2]](diffhunk://#diff-7f7e2a1fff079623a40926984dc95d7de0fac5a9575881e8ec09ce309903a574L197-R197) [[3]](diffhunk://#diff-95e0d5df4894ae4a99b5ad4a37811c64ec30d82af3b01035561976afc3c4293eL119-R125) [[4]](diffhunk://#diff-d567696a32088b29da11cf620c9180ca5000b98120fca1f7ef01d66836109d02L82-R82) [[5]](diffhunk://#diff-2f4d6afbe4caeb1b1438089f2fe978bdfb460346753b5559df20560b6b68ef9dL71-R71)

Test updates:
* Modified tests in `PostingsReport` to use the new `readJournal''` function (`hledger-lib/Hledger/Reports/PostingsReport.hs`). [[1]](diffhunk://#diff-0d9c798b1335019b9475e284450700481b3ce00c8f416523a5f73e83badcc112L304-R304) [[2]](diffhunk://#diff-0d9c798b1335019b9475e284450700481b3ce00c8f416523a5f73e83badcc112L316-R316) [[3]](diffhunk://#diff-0d9c798b1335019b9475e284450700481b3ce00c8f416523a5f73e83badcc112L328-R328)

Miscellaneous:
* Added new language pragmas and updated the module exports to include new functions (`hledger-lib/Hledger/Utils/IO.hs`). [[1]](diffhunk://#diff-5692bda46ff6e1e83272e0e8d71eb87ac9389f0c3695151b6c8370f5d7f42e55R7-R9) [[2]](diffhunk://#diff-5692bda46ff6e1e83272e0e8d71eb87ac9389f0c3695151b6c8370f5d7f42e55R36-R44)

These changes improve the flexibility and robustness of the journal reading functions by allowing them to handle different input sources and text encodings.

## SM summary/doc draft

Previously, hledger could read CSV files containing non-ascii characters only if they are UTF8-encoded.
Now there is a new CSV rule, `encoding ENCODING`, which allows reading CSV files with other encodings.
The following encodings are supported:
...

## Actions

- Let's squash these commits into one when you get a chance

- Let's update the message. I usually don't bother with markdown hyperlinks in these - Github doesn't render them.
  It could be something like

    feat: Implement encoding in rules file

    This uses the encoding library, which supports fewer encodings than text-icu
    but is more portable, not requiring a C library.

    ...

- Let's clarify a bit the user-facing changes as well as the developer-facing ones.
  So far we have this, which is more temporary PR discussion than informative permanent commit message.
  Happy to suggest ideas once I understand this PR myself.

    I also changed the `rReadFn` to get a `Handle` rather than a `Text`.
    Such that we can use non standard functions for reading it.
    This is kind of sideways, since the one place we care about it,
    we don't use it at all. Though it still caused issues, since the input
    wasn't parseable as text.
    I think not using the handle has something to do with how `-` is treated
    as stdin but I didn't do any further tests on that, since this is out of
    the scope of this PR.

simonmichael commented:
Thanks @jokesper.
Now it's time for more review, and I'd welcome help from other reviewers.
Here's some relevant info:

26 files changed

encoding >0.9 dependency added

## Main commit description

    feat: Implement encoding in rules file

    Instead of `text-icu` as [recommended by `text`](https://hackage.haskell.org/package/text-2.1.2/docs/Data-Text-Encoding.html):
    
    > To gain access to a much larger family of encodings, use the `text-icu` package.
    
    we use `encoding`, since `text-icu` requires an external library.
    `encoding` does require a custom setup, which is currently
    affected by a [bug in Cabal](https://github.com/haskell/cabal/issues/6505)
    but it doesn't require an external runtime dependency, which makes it
    easier to distribute on macos and windows.
    
    I also changed the `rReadFn` to get a `Handle` rather than a `Text`.
    Such that we can use non standard functions for reading it.
    This is kind of sideways, since the one place we care about it,
    we don't use it at all. Though it still caused issues, since the input
    wasn't parseable as text.
    I think not using the handle has something to do with how `-` is treated
    as stdin but I didn't do any further tests on that, since this is out of
    the scope of this PR.


## Github AI summary

This pull request includes several changes to the `hledger-lib` library, primarily focusing on modifying the journal reading functions to use `Handle` instead of `Text` and adding support for handling different text encodings. The changes span multiple files and involve updates to function signatures, imports, and helper functions.

Updates to journal reading functions:
* Changed `readJournal` and related functions to accept a `Handle` instead of `Text` for reading journal data (`hledger-lib/Hledger/Read.hs`). [[1]](diffhunk://#diff-bdf99fb149929f11680d239ef2f7508759abea3b651393dc6806a31e1e71e511L222-R227) [[2]](diffhunk://#diff-bdf99fb149929f11680d239ef2f7508759abea3b651393dc6806a31e1e71e511L316-R324)
* Added a new function `readJournal''` that allows reading from `Text` by converting it to a `Handle` (`hledger-lib/Hledger/Read.hs`).

Import and helper function updates:
* Updated imports to include `Handle` and other necessary modules (`hledger-lib/Hledger/Read.hs`, `hledger-lib/Hledger/Read/Common.hs`, `hledger-lib/Hledger/Read/CsvReader.hs`, `hledger-lib/Hledger/Read/RulesReader.hs`). [[1]](diffhunk://#diff-bdf99fb149929f11680d239ef2f7508759abea3b651393dc6806a31e1e71e511L148-R149) [[2]](diffhunk://#diff-fcee3c006b1d2e35b8c279c9daa9963ecdb5910bdc90099f7b3a37b8b529a65dR152) [[3]](diffhunk://#diff-52f2974c62ce69a30e2bddf4d8b6008441715fde0a9d0a4c3ea8c072fa465223L31-R31) [[4]](diffhunk://#diff-95e0d5df4894ae4a99b5ad4a37811c64ec30d82af3b01035561976afc3c4293eR73)
* Added helper functions such as `handleReadFnToTextReadFn` to facilitate the transition from `Text` to `Handle` (`hledger-lib/Hledger/Read/Common.hs`).

Changes in specific readers:
* Updated `CsvReader`, `JournalReader`, `RulesReader`, `TimeclockReader`, and `TimedotReader` to use the new `Handle`-based functions (`hledger-lib/Hledger/Read/CsvReader.hs`, `hledger-lib/Hledger/Read/JournalReader.hs`, `hledger-lib/Hledger/Read/RulesReader.hs`, `hledger-lib/Hledger/Read/TimeclockReader.hs`, `hledger-lib/Hledger/Read/TimedotReader.hs`). [[1]](diffhunk://#diff-52f2974c62ce69a30e2bddf4d8b6008441715fde0a9d0a4c3ea8c072fa465223L57-R68) [[2]](diffhunk://#diff-7f7e2a1fff079623a40926984dc95d7de0fac5a9575881e8ec09ce309903a574L197-R197) [[3]](diffhunk://#diff-95e0d5df4894ae4a99b5ad4a37811c64ec30d82af3b01035561976afc3c4293eL119-R125) [[4]](diffhunk://#diff-d567696a32088b29da11cf620c9180ca5000b98120fca1f7ef01d66836109d02L82-R82) [[5]](diffhunk://#diff-2f4d6afbe4caeb1b1438089f2fe978bdfb460346753b5559df20560b6b68ef9dL71-R71)

Test updates:
* Modified tests in `PostingsReport` to use the new `readJournal''` function (`hledger-lib/Hledger/Reports/PostingsReport.hs`). [[1]](diffhunk://#diff-0d9c798b1335019b9475e284450700481b3ce00c8f416523a5f73e83badcc112L304-R304) [[2]](diffhunk://#diff-0d9c798b1335019b9475e284450700481b3ce00c8f416523a5f73e83badcc112L316-R316) [[3]](diffhunk://#diff-0d9c798b1335019b9475e284450700481b3ce00c8f416523a5f73e83badcc112L328-R328)

Miscellaneous:
* Added new language pragmas and updated the module exports to include new functions (`hledger-lib/Hledger/Utils/IO.hs`). [[1]](diffhunk://#diff-5692bda46ff6e1e83272e0e8d71eb87ac9389f0c3695151b6c8370f5d7f42e55R7-R9) [[2]](diffhunk://#diff-5692bda46ff6e1e83272e0e8d71eb87ac9389f0c3695151b6c8370f5d7f42e55R36-R44)

These changes improve the flexibility and robustness of the journal reading functions by allowing them to handle different input sources and text encodings.

## SM summary/doc draft

Previously, hledger could read CSV files containing non-ascii characters only if they are UTF8-encoded.
Now there is a new CSV rule, `encoding ENCODING`, which allows reading CSV files with other encodings.
The following encodings are supported:
...

## Actions

- Let's squash these commits into one when you get a chance

- Let's update the message. I usually don't bother with markdown hyperlinks in these - Github doesn't render them.
  It could be something like

    feat: Implement encoding in rules file

    This uses the encoding library, which supports fewer encodings than text-icu
    but is more portable, not requiring a C library.

    ...

- Let's clarify a bit the user-facing changes as well as the developer-facing ones.
  So far we have this, which is more temporary PR discussion than informative permanent commit message.
  Happy to suggest ideas once I understand this PR myself.

    I also changed the `rReadFn` to get a `Handle` rather than a `Text`.
    Such that we can use non standard functions for reading it.
    This is kind of sideways, since the one place we care about it,
    we don't use it at all. Though it still caused issues, since the input
    wasn't parseable as text.
    I think not using the handle has something to do with how `-` is treated
    as stdin but I didn't do any further tests on that, since this is out of
    the scope of this PR.

jokesper commented:
Yes, this is due to the interface of `hGetContents` from `encoding` using an implicit parameter.

simonmichael commented:
Edited my last comment a little (re commit message).

I see. I have used ImplicitParams in the past and don't like it much, it always confuses me. But it seems a requirement here.. maybe a benefit. Do you have any insight into what it does for us here ?

simonmichael commented:
I see:

https://hackage.haskell.org/package/encoding-0.9/docs/System-IO-Encoding.html
> By using implicit parameters, it can be used almost as a drop-in replacement

An encoding can be assigned to `?enc`, which works like a global for the System.IO.Encoding functions, instead of needing to provide it as an extra argument.

jokesper commented:
Alternatively we could inline the body of `hGetContents`. It's just reading a lazy bytestring and decoding it.

simonmichael commented:
I kicked off the multi-platform CI builds. encoding is not building on Windows unfortunately, because of:
```
encoding                     > Configuring encoding-0.9...
encoding                     > Error: [Cabal-4345]
encoding                     > Missing dependency on a foreign library:
encoding                     > * Missing (or bad) header file: system_encoding.h
encoding                     > If the header file does exist, it may contain errors that are caught by the C compiler at the preprocessing stage. In this case you can re-run configure with the verbosity flag -v3 to see the error messages.
```

simonmichael commented:
I opened https://github.com/dmwit/encoding/issues/26

jokesper committed:
dev: Fixed windows build

jokesper commented:
I applied the workaround. Though I don't know whether this fixes the issue for cabal on windows. Is this (cabal+windows) even part of our target developer base?

simonmichael commented:
Yes, being easy to build with stack or cabal on any platform is a hledger requirement. Better if we can fix this upstream if possible. 

Here's another note for the doc:

The following encodings are supported (case-insensitive, can also be written with inner spaces or hyphens):

ASCII
UTF8
UTF16
UTF32
ISO88591
ISO88592
ISO88593
ISO88594
ISO88595
ISO88596
ISO88597
ISO88598
ISO88599
ISO885910
ISO885911
ISO885913
ISO885914
ISO885915
ISO885916
CP1250
CP1251
CP1252
CP1253
CP1254
CP1255
CP1256
CP1257
CP1258
KOI8R
KOI8U
GB18030
MacOSRoman
JISX0201
JISX0208
ISO2022JP
ShiftJIS
CP437
CP737
CP775
CP850
CP852
CP855
CP857
CP860
CP861
CP862
CP863
CP864
CP865
CP866
CP869
CP874
CP932

jokesper committed:
;doc: List valid encodings

jokesper commented:
I added the list to the manual. With a link to the cases for the alternative spellings.
though I don't know whether the either is necessary, as the intention of this function is to allow any spelling for an encoding one may have choosen to be work.

simonmichael commented:
@jokesper, beyond the scope of this PR but do you have any thoughts about wider support for text encodings - would we ideally support this for all input files ? Output files ? Or is it best left to external tools, with CSV reading being a special case because we are more often dealing with CSV from diverse sources ?

jokesper commented:
I think having it just for CSVs is fine, since this is the common use case. It's unlikely to get a journal file in a different encoding, though what would be possible, since we changed from `Text` to `Handle` one could maybe implement parsing electronic invoicing. Though this is way out of the scope of HLedger.

Expanding it to other files wouldn't be that hard, except for the design choice of where to document the encoding (command line parameter?). Though in my opinion it's really not needed at all for any other cases.
It doesn't help anyone now, since we effectively already enforced an encoding, and the journal format isn't any common import / export format so no benefit there.

What could be useful is in an csv export, though I don't quite know if this is really relevent. If anyone wants to use it in an CSV export please open an issue.

simonmichael commented:
@jokesper, all: I welcome any review of https://github.com/dmwit/encoding/pull/27, which aims to make encoding build more smoothly on Windows. I'm not quite sure of the breakage risk and how much care is needed.

simonmichael commented:
I have squashed and merged this, plus some doc edits and some test files in examples/. (I didn't get as far as adding automated tests.) I verified that this new feature allows us to successfully convert CSV files with other encodings. Thank you @jokesper !

hseg commented:
Note that the bumping of `encoding` to `0.10` to make use of https://github.com/dmwit/encoding/pull/27 means we shouldn't be disabling the `systemEncoding` since `encoding-0.10` removes it.

i.e. the `.flags.encoding.systemEncoding = false` keys in `stack*.yaml` should
be removed

Stack indeed complains

```
Error: [S-8664]
       Invalid flag specification:
       * Package encoding does not define the following flags (specified in the project-level
         configuration (e.g. stack.yaml)): systemencoding.
         Flags defined by package encoding are: generate-encodings.
```

simonmichael commented:
Thanks! I missed those, now fixed.

simonmichael commented:
Unfortunately, [this change in Common.hs](https://github.com/simonmichael/hledger/pull/2319/files#diff-b1e8037cff372bc60fe965e6d0c52648a764bd12ca3b2fb82db3af832f52e4b8) is causing hledger-web to hang when saving changes to a largeish file (#2389).

simonmichael commented:
A separate thread was needed to avoid deadlock when writing more than a single buffer-ful to the pipe. I have pushed a fix

simonmichael commented:
Just a note that requiring a Handle (or use of inputToHandle) for reading a journal has made hledger-lib a bit harder to use or port, eg to WASM. Was it required as part of journal reading to give a chance to set the encoding ?

jokesper commented:
no, but it was easier to implement as both encoding and text have functions to read from a handle,
but reading into a bytestring and converting (which would be the alternative). Though the alternative shouldn't be to hard.
